### PR TITLE
RTR has multiple threads for assignment of pending tasks now

### LINKED ISSUE #2497: RemoteTaskRunner blocks assignment of tasks on a single slow node
After assigning a task, the RemoteTaskRunner appears to wait for that task to be acknowledged before trying to assign another one, even if it is to a different node.

We had a production issue today where one of our middle managers was having a bad day and not successfully spinning up tasks.  This, quite surprisingly resulted in the overlord totally blocking waiting for that one guy to finish assigning his task even though there were other replicas of the task that could've been assigned and forward progress could've been made.

Specifically, we eventually saw the error from this line of code

https://github.com/druid-io/druid/blob/91f23583f5f624c60c1c099c4fa5b7d120e0bb37/indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java#L713

### PR DESCRIPTION
**Pull request by himanshug**
Fixes #2497

### CODE REVIEW

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java [hunk 1]
```diff
@@ -619,40 +631,54 @@ private boolean tryAssignTask(final Task task, final RemoteTaskRunnerWorkItem ta
...
+          assignedWorker = zkWorkers.get(immutableZkWorker.get().getWorker().getHost());
+          return announceTask(task, assignedWorker, taskRunnerWorkItem);
+        }
+
         log.debug("Worker nodes %s do not have capacity to run any more tasks!", zkWorkers.values());
         return false;
+      } finally {
+        if (assignedWorker != null) {
+          workersWithUnacknowledgedTask.remove(assignedWorker.getWorker().getHost());
+          runPendingTasks();
```
Comment (pjain1) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: why is this needed ?

File: indexing-service/src/main/java/io/druid/indexing/overlord/config/RemoteTaskRunnerConfig.java [hunk 2]
```diff
@@ -48,6 +48,10 @@
   @JsonProperty
   private Period taskShutdownLinkTimeout = new Period("PT1M");
 
+  @JsonProperty
+  @Min(1)
+  private int pendingTasksRunnerNumThreads = 3;
```
Comment (pjain1) [inline indexing-service/src/main/java/io/druid/indexing/overlord/config/RemoteTaskRunnerConfig.java]: should this be documented ?

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java [hunk 3]
```diff
@@ -710,8 +736,8 @@ private boolean announceTask(
         statusLock.wait(waitMs);
         long elapsed = timeoutStopwatch.elapsed(TimeUnit.MILLISECONDS);
         if (elapsed >= waitMs) {
-          log.error(
-              "Something went wrong! [%s] never ran task [%s]! Timeout: (%s >= %s)!",
+          log.makeAlert(
+              "Something went wrong with worker [%s], never ran task [%s]! Timeout: (%s >= %s)!",
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: Generally alert descriptions should be short & simple, as folks have built systems to consolidate alerts with the same description for emails & analysis. So something like `Task assignment timed out` or `Task assignment timed out for worker[%s]` would be good imo, with the additional information (like elapsed & task id) added to the data with `addData`.

Could also do both an alert and a `log.error`.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java [hunk 4]
```diff
@@ -140,11 +138,16 @@
...
+  private final ExecutorService runPendingTasksExec;
 
   // Workers that have been marked as lazy. these workers are not running any tasks and can be terminated safely by the scaling policy.
   private final ConcurrentMap<String, ZkWorker> lazyWorkers = new ConcurrentHashMap<>();
 
+  // workers which were assigned a task and are yet to acknowledge same.
+  // Map: workerId -> taskId
+  private final ConcurrentMap<String, String> workersWithUnacknowledgedTask = new ConcurrentHashMap<>();
+  // tasks which are being tried to be assigned to a worker
+  private final ConcurrentMap<String, String> tryAssignTasks = new ConcurrentHashMap<>();
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: is this taskId -> workerId?

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: oh, no, it's taskId -> taskId. comment would be useful.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java [hunk 5]
```diff
@@ -178,6 +181,7 @@ public RemoteTaskRunner(
     this.workerConfigRef = workerConfigRef;
     this.cleanupExec = MoreExecutors.listeningDecorator(cleanupExec);
     this.resourceManagement = resourceManagement;
+    this.runPendingTasksExec = Execs.multiThreaded(config.getPendingTasksRunnerNumThreads(), "rtr-pending-tasks-runner-%d");
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: the comment "this method uses a single threaded executor" on `runPendingTasks` is no longer true, should be adjusted.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java [hunk 6]
```diff
@@ -547,6 +556,9 @@ public Void call() throws Exception
                   RemoteTaskRunnerWorkItem workItem = pendingTasks.remove(taskId);
                   taskComplete(workItem, null, TaskStatus.failure(taskId));
                 }
+                finally {
+                  tryAssignTasks.remove(taskId);
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: I believe this shouldn't run if `tryAssignTasks.putIfAbsent` returned non-null, as it means some other thread is handling this task.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java [hunk 7]
```diff
@@ -619,40 +631,54 @@ private boolean tryAssignTask(final Task task, final RemoteTaskRunnerWorkItem ta
...
+            == null) {
+          assignedWorker = zkWorkers.get(immutableZkWorker.get().getWorker().getHost());
+          return announceTask(task, assignedWorker, taskRunnerWorkItem);
+        }
+
         log.debug("Worker nodes %s do not have capacity to run any more tasks!", zkWorkers.values());
         return false;
+      } finally {
+        if (assignedWorker != null) {
+          workersWithUnacknowledgedTask.remove(assignedWorker.getWorker().getHost());
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: I believe this shouldn't run if `workersWithUnacknowledgedTask.putIfAbsent` had returned null

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: ah, nm, you already check that.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java [hunk 8]
```diff
@@ -619,40 +631,54 @@ private boolean tryAssignTask(final Task task, final RemoteTaskRunnerWorkItem ta
...
-      } else {
+                )
+            ),
+            task
+        );
+
+        if (immutableZkWorker.isPresent()
+            &&
+            workersWithUnacknowledgedTask.putIfAbsent(immutableZkWorker.get().getWorker().getHost(), task.getId())
+            == null) {
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: I think it's possible for two threads to race to assign to the same worker, and one will succeed and one will fail & we'll get the "Worker nodes %s do not have capacity to run any more tasks" debug message. then runPendingTasks will not be triggered again and so the losing task will not run.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 4 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: will add the comment, ideally I wanted to have a simple ConcurrentSet which does not exist so used ConcurrentMap instead.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 5 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: will update the comment

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 6 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: ah this got messed up in my final touches, will fix

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 8 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: yes, in a race a task might lose, but then whenever a worker frees up we call runPendingTasks() again which should take care of running the task that lost the race.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 1 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: see https://github.com/druid-io/druid/pull/2521/files#r53891406

File: indexing-service/src/main/java/io/druid/indexing/overlord/config/RemoteTaskRunnerConfig.java (hunk 2 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/config/RemoteTaskRunnerConfig.java]: yes, will update the doc.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 4 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: updated

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 5 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: updated

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 6 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: fixed

File: indexing-service/src/main/java/io/druid/indexing/overlord/config/RemoteTaskRunnerConfig.java (hunk 2 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/config/RemoteTaskRunnerConfig.java]: updated

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 3 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: hmmm.. I don't think its that bad, most of the alert msgs are free flowing text, changing it to suggested string does not seem to be buying much. That is, changing it to `Task assignment timed out for worker[%s]` is not that much different.
However, I changed it a bit for clarity.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 1 above)
Comment (pjain1) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: ok

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 8 above)
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: I'm thinking of the case where we have 1 worker and 2 tasks, and runPendingTasks was triggered twice, then I believe this could happen:

- thread 1 assigns task 1 to worker 1
- thread 2 tries to assign task 2 to worker 1, but cannot because it has not yet acknowledged task 1
- now task 2 will not be assigned by either thread, and will stay pending even though there is some capacity

File: docs/content/configuration/indexing-service.md [hunk 9]
```diff
@@ -82,6 +82,7 @@ The following configs only apply if the overlord is running in remote mode:
 |`druid.indexer.runner.maxZnodeBytes`|The maximum size Znode in bytes that can be created in Zookeeper.|524288|
 |`druid.indexer.runner.taskCleanupTimeout`|How long to wait before failing a task after a middle manager is disconnected from Zookeeper.|PT15M|
 |`druid.indexer.runner.taskShutdownLinkTimeout`|How long to wait on a shutdown request to a middle manager before timing out|PT1M|
+|`druid.indexer.runner.pendingTasksRunnerNumThreads`|Number of threads to allocate pending-tasks to workers, must be greater than 1.|3|
```
Comment (gianm) [inline docs/content/configuration/indexing-service.md]: "must be at least 1"? 1 is ok. the number of threads here is sorta the number of slow workers you're allowed before task assignment stops working.

Comment (himanshug) [inline docs/content/configuration/indexing-service.md]: yeah, should be "at least 1"... will update.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 8 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: as soon as worker1 acknowledges (or times out in acknowledging) task1 then runPendingTasks() from line-681 will be triggered again and will assign task2 to worker1. no?

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: ah, I missed that runPendingTasks is triggered after worker assignment. that should be OK then.

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: perhaps a comment there about why it's necessary.

**Comment by gianm** [comment]
lgtm, :+1: after conflict resolution & travis

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 8 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: added comment

**Comment by pjain1** [comment]
:+1:

File: docs/content/configuration/indexing-service.md (hunk 9 above)
Comment (himanshug) [inline docs/content/configuration/indexing-service.md]: updated

**Comment by himanshug** [comment]
fixed conflicts

**Comment by himanshug** [comment]
@gianm @pjain1 can one of you merge this?
